staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_discovery.go

```diff
@@ -40,7 +41,14 @@ func (r *versionDiscoveryHandler) ServeHTTP(w http.ResponseWriter, req *http.Req
 		r.delegate.ServeHTTP(w, req)
 		return
 	}
-	discovery, ok := r.getDiscovery(schema.GroupVersion{Group: pathParts[1], Version: pathParts[2]})
+
+	clusterName := ""
+	cluster := genericapirequest.ClusterFrom(req.Context())
```

Comment:
You're doing this so often you might want to just create genericapirequest.ClusterNameOrEmptyStringFrom(req.Context())

although that's a code smell because it's almost a certainty we'll have a bug in the long run where you do not want to risk an empty string.  If we actually did land this in tree, we'd probably always have a default cluster name, so we would ultimately have genericapirequest.ClusterNameFrom().

Since you've changed this code to always require a cluster name in the "hack" version, you should probably just start with `name, ok := genericapirequest.ClusterNameFrom(ctx)` and if `ok` is not empty you error out right away (wildcards).

Reply:
Completely agree with the proposal here. As you can imaging I built this PR in an iterative way, and yes finally it might even be consistent to error when the clusterName is not set.